server/src/main/java/org/elasticsearch/repositories/IndexSnapshotsService.java

```diff
@@ -0,0 +1,172 @@
+
+            if (snapshotInfo == null || snapshotInfo.state() != SnapshotState.SUCCESS) {
+                // We couldn't find a valid candidate
+                listener.onResponse(Optional.empty());
+                return;
+            }
+
+            // TODO: This is executed in SNAPSHOT_META thread pool and we perform a few blocking operations here, should we execute this
+            // step somewhere else?
+
+            // We fetch BlobStoreIndexShardSnapshots instead of BlobStoreIndexShardSnapshot in order to get the shardStateId that
+            // allows us to tell whether or not this shard had in-flight operations while the snapshot was taken.
+            final BlobStoreIndexShardSnapshots blobStoreIndexShardSnapshots = fetchSnapshotContext.getBlobStoreIndexShardSnapshots();
+            final String indexMetadataId = fetchSnapshotContext.getIndexMetadataId();
+
+            final Optional<ShardSnapshotInfo> indexShardSnapshotInfo = blobStoreIndexShardSnapshots.snapshots()
+                .stream()
+                .filter(snapshotFiles -> snapshotFiles.snapshot().equals(snapshotInfo.snapshotId().getName()))
+                .findFirst()
+                .map(snapshotFiles -> fetchSnapshotContext.createIndexShardSnapshotInfo(indexMetadataId, snapshotFiles));
+
+            listener.onResponse(indexShardSnapshotInfo);
+        }, listener::onFailure);
+
+        repository.getRepositoryData(repositoryDataStepListener);
+    }
+
+    private Repository getRepository(String repositoryName) {
+        final Map<String, Repository> repositories = repositoriesService.getRepositories();
+        return repositories.get(repositoryName);
+    }
+
+    private static class FetchShardSnapshotContext {
+        private final Repository repository;
+        private final RepositoryData repositoryData;
+        private final IndexId indexId;
+        private final ShardId shardId;
+        private final SnapshotInfo snapshotInfo;
+
+        FetchShardSnapshotContext(
+            Repository repository,
+            RepositoryData repositoryData,
+            IndexId indexId,
+            ShardId shardId,
+            SnapshotInfo snapshotInfo
+        ) {
+            this.repository = repository;
+            this.repositoryData = repositoryData;
+            this.indexId = indexId;
+            this.shardId = shardId;
+            this.snapshotInfo = snapshotInfo;
+        }
+
+        private String getIndexMetadataId() throws IOException {
+            final IndexMetaDataGenerations indexMetaDataGenerations = repositoryData.indexMetaDataGenerations();
+            String indexMetadataIdentifier = indexMetaDataGenerations.snapshotIndexMetadataIdentifier(snapshotInfo.snapshotId(), indexId);
+            if (indexMetadataIdentifier != null) {
+                return indexMetadataIdentifier;
+            }
+            // Fallback to load IndexMetadata from the repository and compute the identifier
```

Comment:
Do any of the tests cover this fallback?

Reply:
Yes, we cover that in https://github.com/elastic/elasticsearch/pull/75080/files#diff-a55f8d617afdeb41bb43d0087b7af1b8dbfe2de9840dfa140cc5e749296d721eR117-R129